text/XXX-unwind-FFI.md

```diff
@@ -0,0 +1,286 @@
+* Provide an `unwind` runtime in the standard library
+  that guarantees compatibility with the native exception mechanism
+  provided by the compiler backend.
+* Provide a Cargo option under `[profile.foo]` to specify a `panic` runtime.
+
+[less-unwinding]: https://github.com/rust-lang/rfcs/blob/master/text/1513-less-unwinding.md
+
+# Motivation
+[motivation]: #motivation
+
+This will enable resolving
+[rust-lang/rust#58794](rust-ffi-issue)
+without breaking existing code.
+
+Currently, unwinding through an FFI boundary is always undefined behavior.
+We would like to make the behavior safe
+by aborting when the stack is unwound to an FFI boundary.
+However, there are existing Rust crates
+(notably, wrappers around the `libpng` and `libjpeg` C libraries)
+that rely on the current implementation's compatibility
+with the native exception handling mechanism in
+GCC, LLVM, and MSVC.
+
+This RFC will give crate authors the ability
+to initiate stack-unwinding
+that can propagate through other languages' stack frames,
+as well as tools for ensuring compatibility with those languages.
+
+[rust-ffi-issue]: https://github.com/rust-lang/rust/issues/58794
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Consider an `extern "C"` function that may `panic`:
+
+```rust
+extern "C" fn may_panic(i: i32) {
+    if i < 0 {
+        panic!("Oops, I should have used u32.");
+    }
+}
+```
+
+In a future (TBD) version of Rust,
+calling this function with an argument less than
+zero will cause the program to be aborted.
+This is the only way the compiler can ensure that the function is safe,
+because there is no way for it to know whether or not the calling code
+supports the same implementation of stack-unwinding used for Rust.
+
+As a concrete example,
+this function may be invoked from C code on Linux or BSD
+compiled without support for native stack-unwinding.
+The C runtime would lack the necessary metadata
+to properly propagate the unwinding operation,
+so it would be undefined behavior to let the runtime attempt
+to unwind the C stack frames.
+
+However, if the C code is compiled
+with support for the same stack-unwinding mechanism
+used to by the Rust code,
+unwinding across the FFI boundary is well-defined and safe,
+and in fact it can be a useful way to handle errors
+when working with certain C libraries, such as `libjpeg`.
+
+Thus, the `may_panic` function may be makred `#[unwind(allow)]`,
+which ensures that the unwinding operation will be propagated to the caller
+rather than aborting the process.
+This annotation will also prevent the compiler
+from marking the function as `noexcept`
+(which permits the backend toolchain from optimizing based on the assumption
+that an unwinding operation cannot escape from a function).
+
+This annotation can only be used with an `unsafe` function,
+since the compiler is unable to guarantee
+that the caller will be compiled with a compatible stack-unwinding mechanism.
```

Comment:
`extern "C"` functions can be also called from Rust as well, and there is no guarantee that they will ever be called from C at all. 

AFAICT, if they were to be called from C, it would be up-to-C to mark them unsafe, and make sure that all proof-obligations are satisfied before calling them. 

So I'm not sure what the point of requiring `unsafe` here is.

Reply:
My thought process was that every piece of code imposing a safety-related contract that cannot be checked by `rustc` should be marked `unsafe`. Permitting `unwind` out of an `extern` function imposes such a contract: the calling code must be able to catch or propagate the `unwind` safely.

Reply:
...though I hadn't thought about the inconvenience of statically linking such a function against other Rust code, which would be perfectly safe but would require an `unsafe` block anyway.

Reply:
> My thought process was that every piece of code imposing a safety-related contract that cannot be checked by rustc should be marked unsafe

I'm not sure I follow your train of thought. What makes:

```rust
pub #[unwind(allow)] extern "Rust" fn foo() { ... }
```

special when compared with 

```rust
pub fn bar() { ... }
```

? 

Both can panic, and both can be called from other Rust crates. Why does one need to be unsafe, but the other does not ?

Reply:
If I understand correctly, only the `extern` version can be called from a dynamically-loaded shared library. Is it possible to call `pub fn bar()` in this way? If so, then, since Rust permits custom `panic` runtimes, it should be possible to have undefined behavior by calling `bar` by compiling the shared library and the executable itself with different `panic` runtimes. If that's the case...I'm not sure what the solution would be.